usr.bin/vgafont/vgafont.1
.\" Copyright (c) 1993 Berkeley Software Design, Inc. All rights reserved.
.\" The Berkeley Software Design Inc. software License Agreement specifies
.\" the terms and conditions for redistribution.
.\"	BSDI $Id: vgafont.1,v 1.1.1.1 1993/03/09 15:13:03 polk Exp $
.Dd March 2, 1993
.Dt VGAFONT 1
.Os
.Sh NAME
.Nm vgafont
.Nd load EGA/VGA fonts
.Sh SYNOPSIS
.Nm vgafont
.Op Ar fontname
.Sh DESCRIPTION
.Nm Vgafont
is used to load and restore text-mode fonts on EGA/VGA display adapters.
By default
.Nm vgafont
loads
a VGA font named
.Pa default ;
this may be overridden with the environment variable
.Ev VGAFONT .
.Pp
If
.Ar fontname
does not contain a
.Dq /
character,
.Nm vgafont
will look for the font file relative to the directory
.Pa /usr/share/vgafont ;
otherwise it will treat
.Ar fontname
as a pathname.
.Pp
The font file should
contain character bitmap lines of the form:
.Bd -literal -offset indent
<char-code>: <line1> <line2> ... <lineN>
.Ed
.Pp
where
.Ar char-code
is the hexadecimal code of a character, and
.Ar line1 ,
.Ar line2
and so on
are hexadecimal bitmaps for the character scan lines.
The least significant bit corresponds to the left side,
and scan lines are listed from top to bottom.
The file may contain comments.
Comments start with a
.Dq #
character and run to the end of the line.
Blank lines or lines containing only a comment are ignored.
.Pp
The default VGA character cell size is 8 columns by
16 lines; dots in the 8th column can be duplicated in the
9th column by hardware.
EGA and CGA/MDA cell sizes are 8x14 and 8x8 correspondingly.
If a character definition string contains fewer line octets
than are used by the video adapter, the lower part of the cell
will be loaded with zeroes.
Extra line octets are ignored.
.Sh ENVIRONMENT
If the following environment variable exists, it is utilized by
.Nm vgafont .
.Bl -tag -width Fl
.It Ev VGAFONT
Contains the name of a font which will be used instead of
.Pa default .
.El
.Sh FILES
.Bl -tag -width /usr/share/vgafont -compact
.It Pa /usr/share/vgafont
directory for the font files
.It Pa /dev/vga
VGA device file
.El
.Sh SEE ALSO
.Xr pccons 4 ,
.Xr environ 7
.Sh HISTORY
The
.Nm vgafont
command is
.Ud .


usr.bin/vgafont/vgafont.c
/*
 * Copyright (c) 1993 Berkeley Software Design, Inc. All rights reserved.
 * The Berkeley Software Design Inc. software License Agreement specifies
 * the terms and conditions for redistribution.
 */

static const char rcsid[] = "$Id: vgafont.c,v 1.1.1.1 1993/03/09 15:13:03 polk Exp $";
static const char copyright[] = "Copyright (c) 1993 Berkeley Software Design, Inc.";

/*
 * Load font to EGA or VGA
 *
 * Format of the font file:
 *
 *      # comment
 *      hex-char-code: hex-byte hex-byte hex-byte...
 */

#include <stdio.h>
#include <sys/cdefs.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <signal.h>
#include "vgaioctl.h"
#include "pathnames.h"

void setvga __P((void));
void resetvga __P((void));
int  inb __P((unsigned));
void outb __P((unsigned, int));
int  gethex __P((char **));

int   vgafd;            /* vga file descriptor */
char *vgabase;          /* vga base memory address */

/*
 * The main routine
 */
main(ac, av)
	int ac;
	char **av;
{
	FILE *ff;
	char iline[256];
	char fn[256];
	int lineno;
	char *p;
	char *pl;
	int ccode, nbytes;
	char *getenv(), *index();
	char *fontname = NULL;

	if (ac > 2) {
		fprintf(stderr, "Usage: vgafont [fontname]\n");
		exit(1);
	}

	if (ac == 2)
		fontname = av[1];
	else {
		fontname = getenv("VGAFONT");
		if (fontname == NULL)
			fontname = "default";
	}

	if (index(fontname, '/') == NULL) {
		(void) snprintf(fn, sizeof fn, "%s/%s",
				_PATH_VGAFONT, fontname);
		fontname = fn;
	}

	if( (ff = fopen(fontname, "r")) == NULL ) {
		fprintf(stderr, "vgafont: cannot open font file %s\n", fontname);
		exit(1);
	}

	signal(SIGHUP, SIG_IGN);
	signal(SIGINT, SIG_IGN);
	signal(SIGQUIT, SIG_IGN);
	signal(SIGTERM, SIG_IGN);

	setvga();

	/*
	 * Read the font file
	 */
	lineno = 0;
	while (fgets(iline, sizeof iline, ff) != NULL) {
		lineno++;
		p = iline;
		ccode = gethex(&p);
		if (ccode == -1) {
			while (*p == ' ' || *p == '\t')
				p++;
			if (*p == '#' || *p == '\0' || *p == '\n')
				continue;
		syntax:
			resetvga();
			fprintf(stderr, "vgafont: syntax error: %s, line %d\n",
					fontname, lineno);
			exit(1);
		}
		if (ccode & ~0xff) {
			resetvga();
			fprintf(stderr, "vgafont: bad character code %x: %s, line %d\n",
					ccode, fontname, lineno);
			exit(1);
		}
		while (*p == ' ' || *p == '\t')
			p++;
		if (*p++ != ':')
			goto syntax;

		pl = vgabase + ccode * 32;

		nbytes = 0;
		while ((ccode = gethex(&p)) != -1) {
			if (++nbytes > 32) {
				resetvga();
				fprintf(stderr, "vgafont: too much rows per character: %s, line %d\n",
						fontname, lineno);
				exit(1);
			}
			if (ccode & ~0xff) {
				resetvga();
				fprintf(stderr, "vgafont: too much columns per character: %s, line %d\n",
						fontname, lineno);
				exit(1);
			}
			*pl++ = ccode;
		}
		while (nbytes++ < 32)
			*pl++ = 0;
		while (*p == ' ' || *p == '\t')
			p++;
		if (*p != '#' && *p != '\0' && *p != '\n')
			goto syntax;
	}
	resetvga();
	exit(0);
}

int
inb(port)
	unsigned port;
{
	unsigned char x;

	asm volatile("inb %1, %0" : "=a" (x) : "d" ((short)port));
	return ((int) x);
}

void
outb(port, val)
	unsigned port;
	int val;
{
	asm volatile("outb %0, %1" : : "a" ((char)val), "d" ((short)port));
}

int src_mm;

/*
 * Enable font loading mode
 */
void
setvga()
{
	vgafd = open(_PATH_VGA, 2);
	if (vgafd < 0) {
		fprintf(stderr, "vgafont: cannot open %s\n", _PATH_VGA);
		exit(1);
	}
	vgabase = (char *) 0x8000000;
	if (ioctl(vgafd, VGAIOCMAP, &vgabase) < 0) {
		fprintf(stderr, "vgafont: cannot map VGA memory\n");
		exit(1);
	}

	/* Read GC memory mode register */
	outb(0x3ce, 0x6);
	src_mm = inb(0x3cf);
	if (src_mm == 0xff)
		src_mm = 0xe;   /* EGA, sigh; assume color monitor  */

	/* Relocate VGA memory to 64K at a0000 */
	outb(0x3ce, 0x6);       /* GC Misc register */
	outb(0x3cf, 0x4);

	/* Turn on sequential access to the VRAM */
	outb(0x3c4, 0x4);       /* SEQ Memory Mode register */
	outb(0x3c5, 0x6);

	/* Enable memory plane 2 */
	outb(0x3c4, 0x2);       /* SEQ Plane Enable register */
	outb(0x3c5, 0x4);
}

/*
 * Disable font loading mode
 */
void
resetvga()
{
	/* Enable memory planes 0 and 1 */
	outb(0x3c4, 0x2);       /* SEQ Plane Enable register */
	outb(0x3c5, 0x3);

	/* Turn off sequential access to the VRAM */
	outb(0x3c4, 0x4);       /* SEQ Memory Mode register */
	outb(0x3c5, 0x2);

	/* Relocate VGA memory to 32K at b8000 */
	outb(0x3ce, 0x6);       /* GC Misc register */
	outb(0x3cf, src_mm);

	/* Unmap VRAM */
	(void) ioctl(vgafd, VGAIOCUNMAP, &vgabase);
	close(vgafd);
}

/*
 * Get a hexadecimal number
 */
int
gethex(pp)
	char **pp;
{
	register int n = 0;
	register int c;
	register char *s;

	s = *pp;
	while (*s == ' ' || *s == '\t')
		s++;
	for (;;) {
		c = *s;
		if (c >= '0' && c <= '9')
			n = (c - '0') | (n << 4);
		else if(c >= 'A' && c <= 'F')
			n = (c - 'A' + 10) | (n << 4);
		else if(c >= 'a' && c <= 'f')
			n = (c - 'a' + 10) | (n << 4);
		else {
			if (s == *pp)
				n = -1;
			*pp = s;
			return (n);
		}
		s++;
	}
	/*NOTREACHED*/
}
